# SPAM DETECTION USING NLP

 Download the nltk library

In [1]:
import nltk
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

Import the dataset

In [2]:
import pandas as pd 
import numpy as np 

In [3]:
data=pd.read_csv("C:/Users/91741/Downloads/SMSSpamCollection.tsv",sep="\t",header=None) 
pd.set_option("display.max_colwidth",100)

In [4]:
data.head()

0  \
0   ham   
1  spam   
2   ham   
3   ham   
4   ham   

                                                                                                     1  
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...  
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...  
2                                        Nah I don't think he goes to usf, he lives around here though  
3                        Even my brother is not like to speak with me. They treat me like aids patent.  
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!

Data preprocessing

In [5]:
data.columns=["labels","total text"]

In [6]:
data.head()

labels  \
0    ham   
1   spam   
2    ham   
3    ham   
4    ham   

                                                                                            total text  
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...  
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...  
2                                        Nah I don't think he goes to usf, he lives around here though  
3                        Even my brother is not like to speak with me. They treat me like aids patent.  
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!

In [7]:
data.shape

(5568, 2)

In [8]:
data.labels.value_counts()

ham     4822
spam     746
Name: labels, dtype: int64

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5568 entries, 0 to 5567
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   labels      5568 non-null   object
 1   total text  5568 non-null   object
dtypes: object(2)
memory usage: 87.1+ KB


In [10]:
data.describe()

labels              total text
count    5568                    5568
unique      2                    5165
top       ham  Sorry, I'll call later
freq     4822                      30

In [11]:
data.isnull().sum()

labels        0
total text    0
dtype: int64

In [12]:
data["labels"]=data["labels"].replace("ham","notspam")

REMOVE PUNCTUATION

In [13]:
import string
def punc(text):
    textnonpunc="".join([char for char in text if char not in string.punctuation])
    return textnonpunc
data["clean_text"]=data["total text"].apply(lambda x : punc(x)) 

In [14]:
data.head()

labels  \
0  notspam   
1     spam   
2  notspam   
3  notspam   
4  notspam   

                                                                                            total text  \
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
2                                        Nah I don't think he goes to usf, he lives around here though   
3                        Even my brother is not like to speak with me. They treat me like aids patent.   
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!   

                                                                                            clean_text  
0  Ive been searching for the right words to thank you for this breather I promise i wont take your...  
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive e...  
2                                          Nah I dont think he goes to usf he lives around here though  
3                          Even my brother is not like to speak with me They treat me like aids patent  
4                                                                    I HAVE A DATE ON SUNDAY WITH WILL

Tokenization

In [15]:
import re 

def token(text):
    tokens=re.split('\W+',text)
    return tokens
data["tokenised text"]=data.clean_text.apply(lambda x : token(x.lower()))
data.head()

labels  \
0  notspam   
1     spam   
2  notspam   
3  notspam   
4  notspam   

                                                                                            total text  \
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
2                                        Nah I don't think he goes to usf, he lives around here though   
3                        Even my brother is not like to speak with me. They treat me like aids patent.   
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!   

                                                                                            clean_text  \
0  Ive been searching for the right words to thank you for this breather I promise i wont take your...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive e...   
2                                          Nah I dont think he goes to usf he lives around here though   
3                          Even my brother is not like to speak with me They treat me like aids patent   
4                                                                    I HAVE A DATE ON SUNDAY WITH WILL   

                                                                                        tokenised text  
0  [ive, been, searching, for, the, right, words, to, thank, you, for, this, breather, i, promise, ...  
1  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to...  
2                            [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]  
3         [even, my, brother, is, not, like, to, speak, with, me, they, treat, me, like, aids, patent]  
4                                                           [i, have, a, date, on, sunday, with, will]

Remove Stopwords

In [16]:
from nltk.corpus import stopwords
stop_words=stopwords.words("english")
def remove_stopwords(tokenized_list):
    text = [word for word in tokenized_list if word not in stop_words]
    return text

data['body_text_nostop'] = data['tokenised text'].apply(lambda x: remove_stopwords(x))

data.head()


labels  \
0  notspam   
1     spam   
2  notspam   
3  notspam   
4  notspam   

                                                                                            total text  \
0  I've been searching for the right words to thank you for this breather. I promise i wont take yo...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005. Text FA to 87121 to receive ...   
2                                        Nah I don't think he goes to usf, he lives around here though   
3                        Even my brother is not like to speak with me. They treat me like aids patent.   
4                                                                  I HAVE A DATE ON SUNDAY WITH WILL!!   

                                                                                            clean_text  \
0  Ive been searching for the right words to thank you for this breather I promise i wont take your...   
1  Free entry in 2 a wkly comp to win FA Cup final tkts 21st May 2005 Text FA to 87121 to receive e...   
2                                          Nah I dont think he goes to usf he lives around here though   
3                          Even my brother is not like to speak with me They treat me like aids patent   
4                                                                    I HAVE A DATE ON SUNDAY WITH WILL   

                                                                                        tokenised text  \
0  [ive, been, searching, for, the, right, words, to, thank, you, for, this, breather, i, promise, ...   
1  [free, entry, in, 2, a, wkly, comp, to, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, to...   
2                            [nah, i, dont, think, he, goes, to, usf, he, lives, around, here, though]   
3         [even, my, brother, is, not, like, to, speak, with, me, they, treat, me, like, aids, patent]   
4                                                           [i, have, a, date, on, sunday, with, will]   

                                                                                      body_text_nostop  
0  [ive, searching, right, words, thank, breather, promise, wont, take, help, granted, fulfil, prom...  
1  [free, entry, 2, wkly, comp, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, 87121, receiv...  
2                                                 [nah, dont, think, goes, usf, lives, around, though]  
3                                              [even, brother, like, speak, treat, like, aids, patent]  
4                                                                                       [date, sunday]

In [17]:
 data.drop(columns=["total text",'clean_text','tokenised text'],axis=1,inplace=True)

In [18]:
pd.set_option("display.max_colwidth",200)
data.head()

labels  \
0  notspam   
1     spam   
2  notspam   
3  notspam   
4  notspam   

                                                                                                                                                  body_text_nostop  
0                                 [ive, searching, right, words, thank, breather, promise, wont, take, help, granted, fulfil, promise, wonderful, blessing, times]  
1  [free, entry, 2, wkly, comp, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, 87121, receive, entry, questionstd, txt, ratetcs, apply, 08452810075over18s]  
2                                                                                                             [nah, dont, think, goes, usf, lives, around, though]  
3                                                                                                          [even, brother, like, speak, treat, like, aids, patent]  
4                                                                                                                                                   [date, sunday]

Lemmetization

In [19]:
from nltk.stem import WordNetLemmatizer
wn = WordNetLemmatizer()
def lemmatizing(tokenized_text):
    texts = [wn.lemmatize(char) for char in tokenized_text]
    return texts

data['body_text_lemmatized'] = data['body_text_nostop'].apply(lambda x: lemmatizing(x))
data.drop(columns=["body_text_nostop"],axis=1,inplace=True)
data.head()

labels  \
0  notspam   
1     spam   
2  notspam   
3  notspam   
4  notspam   

                                                                                                                                              body_text_lemmatized  
0                                   [ive, searching, right, word, thank, breather, promise, wont, take, help, granted, fulfil, promise, wonderful, blessing, time]  
1  [free, entry, 2, wkly, comp, win, fa, cup, final, tkts, 21st, may, 2005, text, fa, 87121, receive, entry, questionstd, txt, ratetcs, apply, 08452810075over18s]  
2                                                                                                                [nah, dont, think, go, usf, life, around, though]  
3                                                                                                           [even, brother, like, speak, treat, like, aid, patent]  
4                                                                                                                                                   [date, sunday]

Count Vectorization

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
count_vector=CountVectorizer(analyzer=lemmatizing)
x_counts=count_vector.fit_transform(data["body_text_lemmatized"])

In [21]:
x_counts.shape

(5568, 8912)

In [22]:
feature_names = count_vector.get_feature_names_out()
for feature_name in feature_names:
    print(feature_name)


0
008704050406
0089my
0121
01223585236
01223585334
0125698789
02
020603
0207
02070836089
02072069400
02073162414
02085076972
020903
021
050703
0578
06
060505
061104
07008009200
07046744435
07090201529
07090298926
07099833605
071104
07123456789
0721072
07732584351
07734396839
07742676969
07753741225
0776xxxxxxx
07786200117
077xxx
078
07801543489
07808
07808247860
07808726822
07815296484
07821230901
0784987
0789xxxxxxx
0794674629107880867867
0796xxxxxx
07973788240
07xxxxxxxxx
0800
08000407165
08000776320
08000839402
08000930705
08000938767
08001950382
08002888812
08002986030
08002986906
08002988890
08006344447
0808
08081263000
08081560665
0825
0844
08448350055
08448714184
0845
08450542832
08452810071
08452810073
08452810075over18s
0870
08700621170150p
08701213186
08701237397
08701417012
08701417012150p
0870141701216
087016248
08701752560
087018728737
0870241182716
08702490080
08702840625
08702840625comuk
08704439680
08704439680tscs
08706091795
0870737910216yrs
08707500020
08707509020
08

limit_output extension: Maximum message size of 10000 exceeded with 67599 characters

Data Distribution

In [23]:
x=x_counts.toarray()
y=data.labels

Label Encoding

In [24]:
from sklearn.preprocessing import LabelEncoder
L=LabelEncoder()
y=L.fit_transform(y)

Data splitting

In [25]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(x,y,random_state=0,test_size=0.2)

In [26]:
X_train.shape

(4454, 8912)

In [27]:
X_test.shape

(1114, 8912)

Logistic Regression

In [28]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,Y_train)
Y_pred = model.predict(X_test)
from sklearn.metrics import accuracy_score,classification_report
print("Accuracy Score:",accuracy_score(Y_test,Y_pred))
print("Classification report",classification_report(Y_test,Y_pred))

Accuracy Score: 0.9820466786355476
Classification report               precision    recall  f1-score   support

           0       0.98      1.00      0.99       955
           1       0.99      0.88      0.93       159

    accuracy                           0.98      1114
   macro avg       0.99      0.94      0.96      1114
weighted avg       0.98      0.98      0.98      1114



In [29]:
pd.crosstab(Y_test,Y_pred)

col_0    0    1
row_0          
0      954    1
1       19  140

Decision Tree

In [30]:
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(criterion='entropy',random_state=0)
model.fit(X_train,Y_train)
y_pred = model.predict(X_test)
print("Accuracy Score:",accuracy_score(Y_test,y_pred))
print("Classification report",classification_report(Y_test,y_pred))

Accuracy Score: 0.9694793536804309
Classification report               precision    recall  f1-score   support

           0       0.98      0.99      0.98       955
           1       0.93      0.86      0.89       159

    accuracy                           0.97      1114
   macro avg       0.95      0.92      0.94      1114
weighted avg       0.97      0.97      0.97      1114



In [ ]:
pd.crosstab(Y_test,y_pred)

Random Forest

In [31]:
from sklearn.ensemble import RandomForestClassifier
model2 = RandomForestClassifier(random_state=42)
model2.fit(X_train, Y_train)
y_pred2 = model2.predict(X_test)
print("Accuracy Score:",accuracy_score(Y_test,y_pred2))
print("Classification report",classification_report(Y_test,y_pred2)) 

Accuracy Score: 0.9748653500897666
Classification report               precision    recall  f1-score   support

           0       0.97      1.00      0.99       955
           1       1.00      0.82      0.90       159

    accuracy                           0.97      1114
   macro avg       0.99      0.91      0.94      1114
weighted avg       0.98      0.97      0.97      1114



In [ ]:
pd.crosstab(Y_test,y_pred2)

So from above machine learning analysis we conclude that Logistic Regression with 98% is best for Spam detection